src/config.ts

```diff
@@ -1,26 +1,48 @@
 import "dotenv/config";
-import { cleanEnv, json, num, str } from "envalid";
-import { PollingOptions } from "grammy";
+import z from "zod";
 
-export const config = cleanEnv(process.env, {
-  NODE_ENV: str({ choices: ["development", "production"] }),
-  LOG_LEVEL: str({
-    choices: ["trace", "debug", "info", "warn", "error", "fatal", "silent"],
-  }),
-  DATABASE_URL: str(),
-  REDIS_URL: str(),
-  BOT_SERVER_HOST: str({
-    default: "0.0.0.0",
-  }),
-  BOT_SERVER_PORT: num({
-    default: 80,
-  }),
-  BOT_ALLOWED_UPDATES: json<PollingOptions["allowed_updates"]>({
-    default: [],
-  }),
-  BOT_TOKEN: str(),
-  BOT_WEBHOOK: str(),
-  BOT_ADMIN_USER_ID: num(),
+const updates = [
+  "message",
+  "poll",
+  "poll_answer",
+  "my_chat_member",
+  "chat_member",
+  "chat_join_request",
+  "edited_message",
+  "channel_post",
+  "edited_channel_post",
+  "inline_query",
+  "chosen_inline_result",
+  "callback_query",
+  "shipping_query",
+  "pre_checkout_query",
+] as const;
+
+const configSchema = z.object({
+  NODE_ENV: z.enum(["development", "production"]),
+  LOG_LEVEL: z.enum([
+    "trace",
+    "debug",
+    "info",
+    "warn",
+    "error",
+    "fatal",
+    "silent",
+  ]),
+  DATABASE_URL: z.string(),
+  REDIS_URL: z.string(),
+  BOT_SERVER_HOST: z.string().default("0.0.0.0"),
+  BOT_SERVER_PORT: z.coerce.number().positive().default(80),
+  BOT_ALLOWED_UPDATES: z.preprocess(
+    (v: unknown) => JSON.parse(String(v)),
+    z.array(z.enum(updates))
+  ),
+  BOT_TOKEN: z.string(),
+  BOT_WEBHOOK: z.string().url(),
+  BOT_ADMIN_USER_ID: z.coerce.number().finite(),
 });
 
-export type Config = typeof config;
+export const config = configSchema.parse(process.env);
+export const isDev = config.NODE_ENV === "development";
+export const isProd = config.NODE_ENV === "production";
```

Comment:
Can we have these variables inside the config schema?

```suggestion
  isDev: z.boolean(),
  isProd: z.boolean(),
});

export const config = configSchema.parse({
  ...process.env,
  isDev: process.env.NODE_ENV === "development",
  isProd: process.env.NODE_ENV === "production",
})
```

Reply:
How about adding them inside a parseConfig function?
```typescript
const parseConfig = (env: NodeJS.ProcessEnv) => {
  const config = configSchema.parse(env);

  return {
    ...config,
    isDev: config.NODE_ENV === "development",
    isProd: config.NODE_ENV === "production",
  };
};

export const config = parseConfig(process.env);
```

Reply:
Looks good to me